src/php/Command/ReplCommand.php

```diff
@@ -21,102 +20,123 @@ final class ReplCommand
-        $this->io->output($this->style->yellow("Welcome to the Phel Repl\n"));
-        $this->io->output('Type "exit" or press Ctrl-D to exit.' . "\n");
+        $this->io->output($this->style->yellow('Welcome to the Phel Repl' . PHP_EOL));
+        $this->io->output('Type "exit" or press Ctrl-D to exit.' . PHP_EOL);
 
         $this->loopReadLineAndAnalyze();
     }
 
-    /**
-     * @throws ExitException
-     */
     private function loopReadLineAndAnalyze(): void
     {
         while (true) {
             try {
-                if ($this->io->isBracketedPasteSupported()) {
-                    $this->io->output(self::ENABLE_BRACKETED_PASTE);
-                }
-                $input = $this->io->readline(self::PROMPT);
-                if ($this->io->isBracketedPasteSupported()) {
-                    $this->io->output(self::DISABLE_BRACKETED_PASTE);
-                }
-
-                try {
-                    $this->checkInputAndAnalyze($input);
-                } catch (ExitException $e) {
-                    break;
-                }
+                $this->addLineFromPromptToBuffer();
+                $this->checkExitInputBuffer();
+                $this->analyzeInputBuffer();
+            } catch (ExitException $e) {
+                break;
             } catch (Throwable $e) {
-                $this->io->output($e->getTraceAsString() . PHP_EOL);
+                $this->inputBuffer = [];
+                $this->io->output($this->style->red($e->getMessage() . PHP_EOL));
             }
         }
 
-        $this->io->output($this->style->yellow("Bye!\n"));
+        $this->io->output($this->style->yellow('Bye!' . PHP_EOL));
+    }
+
+    private function addLineFromPromptToBuffer(): void
+    {
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::ENABLE_BRACKETED_PASTE);
+        }
+
+        $prompt = empty($this->inputBuffer) ? self::INITIAL_PROMPT : self::OPEN_PROMPT;
+        $input = $this->io->readline($prompt);
+
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::DISABLE_BRACKETED_PASTE);
+        }
+
+        if ($input === null) {
+            $input = self::EXIT_REPL;
+        }
+
+        $this->inputBuffer[] = $input;
     }
 
     /**
      * @throws ExitException
      */
-    private function checkInputAndAnalyze(?string $input): void
+    private function checkExitInputBuffer(): void
     {
-        if (null === $input || 'exit' === $input) {
+        $firstInput = $this->inputBuffer[0] ?? '';
+
+        if (self::EXIT_REPL === $firstInput) {
             throw new ExitException();
         }
+    }
 
-        if ('' === $input) {
+    private function analyzeInputBuffer(): void
+    {
+        if ('' === end($this->inputBuffer)) {
             return;
         }
 
-        $this->io->addHistory($input);
+        $inputAsString = implode(PHP_EOL, $this->inputBuffer);
+        $this->io->addHistory($inputAsString);
```

Comment:
For the example:
```
>>> (
... +
... 1
... 1)
```

The history contains following entries:
* `(`
* `(\n+`
* `(\n+\n1`
* `(\n+\n1\n1)`

Is that behavior correct? (Sorry not tested)

Reply:
Right, it shouldn't be like that. I will change it to add to the history ONLY after being successfully parsed.
In your example:
```
>>> (
... +
... 1
... 1)
```
it should store in the history only one line:
```
(\n+\n1\n1)
```

Reply:
Nope, with is correct. We store the plain string as 
```php
$input = implode(PHP_EOL, $this->inputBuffer); // -> (\n+\n1\n1)
```